### PR TITLE
[GEOS-11324] MapML vector style classes

### PR DESCRIPTION
**Pull request by turingtestfail**
# Checklist

- [ ] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [ ] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [ ] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [ ] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [ ] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [ ] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by aaime** [comment]
@prushforth this PR attempts to translate SLDs into CSS and apply it to the geometries, however, for what I've tried, we have little to play with. Basic line styles are working, the color of point symbolizes goes through, but for example the fill of polygon does not seem to be applied, and using fill or background-color does not seem to make a difference.

To test, set a layer in "features" mode, without tiling.

Do you have a list of CSS properties the client can work with?

**Comment by prushforth** [comment]
> Do you have a list of CSS properties the client can work with?

Not really - we are copying classes from the map-feature onto the svg \<g> and child \<path> that renders it, and including styles in the preceding sibling element as actual \<style> elements.  The new viewer will work better, I hope.  I will test it later today (I'm at work right now and unable to build geoserver or do much useful work beyond email).  It's going to be a case of updating the viewer to work with the new GeoServer capabilities in all likelihood, as we haven't had a ready source of vector data before this work.

I will submit a PR to GeoServer main today that I hope to have finished by Friday, but which contains the [0.13.x version](https://www.npmjs.com/package/@maps4html/web-map-custom-element/v/0.13.0) of the viewer.

**Comment by prushforth** [comment]
Currently, the PR is adding class values to the root geometry type element: 

image

If you put the class on the \<map-feature> instead, it will apply the styles you're sending:

image

**Comment by prushforth** [comment]
I apologize that our documentation of this feature is not up to date - that's my fault.  It's also an area that we haven't had much call to work on, but having a ready source of styled vector content will definitely help motivate us and will shape the viewer's future capabilities.

**Comment by turingtestfail** [comment]
> I apologize that our documentation of this feature is not up to date - that's my fault. It's also an area that we haven't had much call to work on, but having a ready source of styled vector content will definitely help motivate us and will shape the viewer's future capabilities.

No worries at all.  It was straightforward to move the style class identifier up to the feature.  The latest push is ready for your testing.

**Comment by turingtestfail** [comment]
> > I apologize that our documentation of this feature is not up to date - that's my fault. It's also an area that we haven't had much call to work on, but having a ready source of styled vector content will definitely help motivate us and will shape the viewer's future capabilities.
> 
> No worries at all. It was straightforward to move the style class identifier up to the feature. The latest push is ready for your testing.

Also, please note that there was a naming collision in Feature - the featuretype was being written to a "clazz" attribute that was aliased to "class".  I changed that alias to "featuretype-class".

**Comment by prushforth** [comment]
Looking pretty good now - note that the viewer has hard-coded the svg stroke- attributes and this is causing some competition with the \<LineSymbolizer> rules

image

I will see if I can fix that asap.

**Comment by prushforth** [comment]
I believe the features are being duplicated - once for each symbolizer perhaps?: 

image

**Comment by aaime** [comment]
That's by design, in SLD each symbolizer paints on top of the previous one, to generate visual effects like cased roads, complex symbols and the like. Can this be achieved without duplicating features?

**Comment by prushforth** [comment]
I think so, just put the symbolizers in the class attribute in order of rendering?

**Comment by turingtestfail** [comment]
Correct.  It currently generates a feature for every symbolizer that is
applicable.

**Comment by aaime** [comment]
> I think so, just put the symbolizers in the class attribute in order of rendering?

That's surprising... so I can put a stroke twice, with two different colors and two different sizes, to get an effect like this one:

image

In CSS that's not possible, the second stroke definition would overwrite the first. Is SVG working differently?

**Comment by prushforth** [comment]
> Can this be achieved without duplicating features?

> Is SVG working differently?

Our renderer is generating SVG so its the renderer that is doing the duplication.  Part of the goal of this project is to find the limits of what is possible on the Web, especially in CSS wrt maps and push that as a change request.

**Comment by turingtestfail** [comment]
Should I delimit multiple classes with spaces, similar to CSS?

**Comment by prushforth** [comment]
> Should I delimit multiple classes with spaces, similar to CSS?

Yes, please

**Comment by turingtestfail** [comment]
Done and pushed.  Please note that the latest push will convert the
geometry type to another type if the first symbolizer type does not match
what is found in the featuresource.  Let me know if that is the right
approach or if you want me to remove that functionality and leave the
geometry type as is.

**Comment by turingtestfail** [comment]
Actually, one more question: For the feature representation, should there
be a way to pass scale denominators/zoom layering information to the client
from the SLD Rules? If so where would you want to put that information in
the XML?

**Comment by prushforth** [comment]
> For the feature representation, should there
be a way to pass scale denominators/zoom layering information to the client
from the SLD Rules? If so where would you want to put that information in
the XML?

It's a good question, and I'm not sure I fully understand it. We think CSS will have a major role in deciding (at least, applying specified rules for) the scale / zoom ranges within which a feature should be rendered and with what properties.  However that is getting out beyond the forseeable horizon.

In the meantime, we spent A LOT of time working on the \<map-feature zoom="..." min="..." max="..."> attributes.

The `zoom` attribute says that what the "native" zoom/scale of the feature is.  The `min` and `max` attributes dictate for now what the display scale range of a feature is, at the per-feature level.  When you have a \<map-link rel="features"> it pulls in features at each map movement, and the `zoom` property of the feature, if not actually specified is set from the zoom of the map at the moment the feature joins the fun.  The `min` and `max` values (again, which specify only the visible zoom range), if not specified on the feature as an attribute, are dictated by the \<map-input type=zoom min="..." max="..."> values, or failing that by a fallback search for relevant values up to and including the applicable values from the projection.

So, long story short, we might eventually want to create a \<map-link media="..."> attribute that contains a CSS media query that pertains to the map zoom to allow the viewer to automatically select which \<map-link> to load based on the viewer zoom situation.  In other words, no action is required at this time! I think!

**Comment by aaime** [comment]
It seems this one is ready to become a PR for the official repository. Can you confirm?

**Comment by prushforth** [comment]
Good to go!

**Comment by prushforth** [comment]
Joe - thanks for working on the scale-dependent symbolizers.  I have tried it locally (rebased on main, actually).

The following modified style rules for the states shapefile's population.sld (CBMTILE projection):

```xml
 <Rule>
          <Title>&gt; 4M</Title>
          <!-- like a linesymbolizer but with a fill too -->
          <ogc:Filter>
            <ogc:PropertyIsGreaterThan>
             <ogc:PropertyName>PERSONS</ogc:PropertyName>
             <ogc:Literal>4000000</ogc:Literal>
            </ogc:PropertyIsGreaterThan>
          </ogc:Filter>
          <MinScaleDenominator>2000</MinScaleDenominator>
          <MaxScaleDenominator>5000000</MaxScaleDenominator>          <PolygonSymbolizer>
             <Fill>
                <!-- CssParameters allowed are fill (the color) and fill-opacity -->
                <CssParameter name="fill">#4df7ff</CssParameter>
                <CssParameter name="fill-opacity">0.7</CssParameter>
             </Fill>     
          </PolygonSymbolizer>
        </Rule>
        <Rule>
          <Title>&gt; 4M</Title>
          <!-- like a linesymbolizer but with a fill too -->
          <ogc:Filter>
            <ogc:PropertyIsGreaterThan>
             <ogc:PropertyName>PERSONS</ogc:PropertyName>
             <ogc:Literal>4000000</ogc:Literal>
            </ogc:PropertyIsGreaterThan>
          </ogc:Filter>
          <MinScaleDenominator>5000000</MinScaleDenominator>
                    <PolygonSymbolizer>
             <Fill>
                <!-- CssParameters allowed are fill (the color) and fill-opacity -->
                <CssParameter name="fill">#4D4DFF</CssParameter>
                <CssParameter name="fill-opacity">0.7</CssParameter>
             </Fill>     
          </PolygonSymbolizer>
</Rule>
``` 
works like this in the style previewer:


https://github.com/geosolutions-it/geoserver/assets/2437285/30e4d1b7-63e5-4c6c-8b32-9b121bc4e319

but the max scale denominator does not work quite as it does in the openlayers style preview:

https://github.com/geosolutions-it/geoserver/assets/2437285/f8347c66-23a6-4cd2-bceb-41e3f29412f0

**Comment by prushforth** [comment]
If I reverse the order of the rules it seems that it only picks up and applies the first rule.

**Comment by turingtestfail** [comment]
Found the bug.  I was breaking out of the rule loop instead of
continuing to the next iteration (ugggh).  Please pull and try again.

Joe

**Comment by prushforth** [comment]
It works much better now!  Thanks for your quick action.  Probably won't get another look at it until tonight/tomorrow.
Cheers
